README.md
BODR
====

Blue Obelisk Data Repository - your one stop place of element and isotope data.

Download
========

* Debian package: https://packages.debian.org/search?keywords=bodr

How to Cite
===========

If you use this library, please cite:

    Guha, R., Howard, M. T., Hutchison, G. R., Murray-Rust, P., Rzepa, H., Steinbeck, C.,
    Wegner, J., Willighagen, E. L., May 2006. The Blue Obelisk - interoperability in
    chemical informatics. J. Chem. Inf. Model. 46 (3), 991-998.

See http://dx.doi.org/10.1021/ci050400b


bodr/NEWS
Release version 11 SNAPSHOT

        * elements/symbols.bibxml: added two references for recently confirmed elements
        * elements/elements.xml: fixed the element symbol for Lv
        * elements/elements.xml: updated group information
        * elements/elements.xml: update covalent radii per Pyykko 2009

Release version 10 2013-08-23

        * elements/elements.xml: Changed atom IDs of Lv and Fl and their
          name histories. Fixed all atomic masses accordingly to the latest
          publication of the IUPAC.
        * elements/mass.bibxml: Added latest atomic table publication.
        * tests/AtWt.xsl: Fixed handling of non-breaking spaces and the
          latest IUPAC website.
        * isotopes/isotopes.*: Added the half-lives of 209Th and 238Th (fixes #9)
        * elements/elements.xml: Fixed the groups of Ds and Mt
        * elements/elements.xml|symbols.bibxml: Lv and Fl got official names
        * README, COPYING: in agreement with the following email, changed the license to CCZero
        * elements/elements.xml: Fixed description for Copernicium.
        * crystal/space-groups.xml: fixed Hall name for group 205.
        * elements/elements.xml: Fixed atomic weight for Rutherfordium
          accordingly to IUPAC AtWt table.

Release version 9 2010-04-25

	* Atomic weights have been updated to correspond with the IUPAC table (KDE#192551).
	* Minor bug-fixes in the elements database (KDE#192616).
        * Fix in the space group table.
        * Added name for element 112: copernicum (KDE#235313).
        * Updated mass of H2 to current value from webelements.com (closes #2989934 on sf.net)

Release version 8 2008-03-02

        * Uses CML 2.5, with two tweaks: missing Rg and Ds; allow <scalar> inside <isotope>
        * Fixes the distcheck errors: correct dictionary schematron (closes #1905375 on sf.net)

Release version 7 2008-03-01

        * Adds spacegroup information
        * Removed unnecessary paths (closes #1713504 sf.net)
        * Make system fixes
        * Output PUBLIC and SYSTEM IDs for XHTML
        * Adding documentation for discoveryDate
        * Added information about elements 112-118

Release version 6 2007-03-10

        * The .txt source files now mention the license too, instead of '??'
        * updated elemental masses according to a 2006 IUPAC publication
          (closes SF bug #1660527)

Release version 5 2006-12-27

        * The actinoids are now listed as 'other metal'
          (closes SF bug #1603063)
        * The acidity information was removed
          (closes SF bug #1603066)

Release Version 4 2006-10-16

        * Isotope and element data now conforms a Schematron convention.
        * XML files are now valid CML 2.3 files
        * Fixed name of C.W.Scheele 
        * Added element crystal structure data
        * Added country where elements were discovered

Release Version 3 2006-06-07

	* Fixed a stupid mistake in the newly shipped pkgconfig file.
	See Version 2 for other changes.

Release Version 2 2006-06-07 

	* Fixed missing root node in isotopes.xml.
	* Fixed build issues.
	* Added isotope info: decay and spin.
        * Added HTML bibliography for elements.xml

Release Version 1 2006-02-08

	* first tar.gz release



bodr/autogen.sh
#!/bin/sh

## all initial declarations, overwrite them using e.g. 'ACLOCAL=aclocal-1.7 AUTOMAKE=automake-1.7 ./autogen.sh'
ACLOCAL=${ACLOCAL:-aclocal}
AUTOCONF=${AUTOCONF:-autoconf}
AUTOMAKE=${AUTOMAKE:-automake}
#INTLTOOLIZE=${INTLTOOLIZE:-intltoolize}

## check, if all binaries exist and fail with error 1 if not
if [ -z `which $ACLOCAL` ] ; then echo "Error. ACLOCAL=$ACLOCAL not found." >&2 && exit 1 ; fi
if [ -z `which $AUTOCONF` ] ; then echo "Error. AUTOCONF=$AUTOCONF not found." >&2 && exit 1 ; fi
if [ -z `which $AUTOMAKE` ] ; then echo "Error. AUTOMAKE=$AUTOMAKE not found." >&2 && exit 1 ; fi
#if [ -z `which $INTLTOOLIZE` ] ; then echo "Error. INTLTOOLIZE=$INTLTOOLIZE not found." >&2 && exit 1 ; fi

## find where automake is installed and get the version
AUTOMAKE_PATH=${AUTOMAKE_PATH:-`which $AUTOMAKE | sed 's|\/bin\/automake.*||'`}
AUTOMAKE_VERSION=`$AUTOMAKE --version | grep automake | awk '{print $4}' | awk -F. '{print $1"."$2}'`

## automake files we need to have inside our source
if [ $AUTOMAKE_VERSION = "1.7" ] ; then
        AUTOMAKE_FILES="missing mkinstalldirs install-sh"
else
        AUTOMAKE_FILES="missing install-sh"
fi

## our help output - if autogen.sh was called with -h|--help or unkbown option
autogen_help() {
	echo
	echo "./autogen.sh: Produces all files necessary to build the bodr project files."
	echo "              The files are linked by default, if you run it without an option."
	echo
	echo "    -v        Be more verbose about every step (debugging)."
	echo "    -f FILE   Output everything to FILE (debugging). Useful for debug output."
	echo "    -c        Copy files instead to link them."
	echo "    -h        Print this message."
	echo
	echo "  You can overwrite the automatically determined location of aclocal (>= 1.7),"
	echo "  automake (>= 1.7), autoconf using:"
	echo
	echo "    ACLOCAL=/foo/bin/aclocal-1.8 AUTOMAKE=automake-1.8 ./autogen.sh"
	echo
}

## check if $AUTOMAKE_FILES were copied to our source
## link/copy them if not - necessary for e.g. gettext, which seems to always need mkinstalldirs
autogen_if_missing() {
	for file in $AUTOMAKE_FILES ; do
		if [ ! -e "$file" ] ; then
			if [ -n "$VERBOSE" ]; then
				echo "DEBUG: $COPYACTION -f $AUTOMAKE_PATH/share/automake-$AUTOMAKE_VERSION/$file ." >&2
			fi
			$COPYACTION -f $AUTOMAKE_PATH/share/automake-$AUTOMAKE_VERSION/$file .
		fi
	done
}

## link/copy the necessary files to our source to prepare for a build
autogen() {
	#$INTLTOOLIZE $DEBUG -f $COPYOPTION
	$ACLOCAL $VERBOSE
	$AUTOMAKE --gnu -a $VERBOSE $COPYOPTION
	autogen_if_missing
	$AUTOCONF $DEBUG $VERBOSE
}

## the main function
COPYACTION="ln -s"
while getopts "chvf:" options; do
	case "$options" in
		h)
			autogen_help
			exit 0
		;;
		c)
			COPYACTION="cp"
			COPYOPTION="-c"
		;;
		f)
			OUTPUTFILE=$OPTARG
		;;
		v)
			DEBUG="--debug"
			VERBOSE="--verbose"
		;;
	esac
done

if [ -n "$OUTPUTFILE" ]; then
	exec &>$OUTPUTFILE
fi
autogen

## ready to rumble
echo "Run ./configure with the appropriate options, then make and enjoy."

exit 0



bodr/scripts/elements.pl
#!/usr/bin/perl
use strict;
use diagnostics;

my @elements;

# Read in the data
&readColors;
&readNamesAndSymbols;
&readAtomicMasses;
&readExactMasses;
&readIonization;
&readElectronAffinity;
&readEthymology;
&readPauling;
&readRadiiCovalent;
&readRadiiVanDerWaals;
&readDatesAndDiscoveres;
&readPeriods;
&readBlocks;
&readElectronicConfiguration;
&readMeltingpoint;
&readBoilingpoint;
&readGroups;
&readFamilies;
&readCountries;
&readDensity;

print "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n";
print "<list id=\"chemicalElement\" convention=\"bodr:elements\"\n";
print "      title=\"properties of the elements\"\n\n";

print "      xmlns=\"http://www.xml-cml.org/schema\" \n";
print "      xmlns:xml=\"http://www.w3.org/XML/1998/namespace\" \n";
print "      xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" \n";
print "      xmlns:bo=\"http://www.blueobelisk.org/dict/terminology\" \n";
print "      xmlns:boUnits=\"http://www.blueobelisk.org/dict/units\" \n";
print "      xmlns:units=\"http://www.xml-cml.org/units/units\" \n";
print "      xmlns:siUnits=\"http://www.xml-cml.org/units/siUnits\" \n";
print "      xmlns:bibx=\"http://bibtexml.sf.net/\" \n\n";
print "      xsi:schemaLocation=\"http://www.xml-cml.org/schema ../schemas/cml25.xsd\n";
print "                           http://bibtexml.sf.net/       ../schemas/bibtexml.xsd\">\n\n";

print "  <metadataList>\n";
print "    <!-- manually updated -->\n\n";
print "    <metadata name=\"dc:title\" content=\"Blue Obelisk Element Repository\" />\n";
print "    <metadata name=\"dc:creator\" content=\"The Blue Obelisk Movement\" />\n";
print "    <metadata name=\"dc:license\" content=\"The MIT License\" />\n";
print "    <metadata name=\"dc:contributor\" content=\"Geoffrey R. Hutchison\" />\n";
print "    <metadata name=\"dc:contributor\" content=\"Carsten Niehaus\" />\n";
print "    <metadata name=\"dc:contributor\" content=\"Egon Willighagen\" />\n";
print "    <metadata name=\"dc:description\" content=\"Database of elements and elemental properties (names, symbols, masses, exact masses, van der Waals radii, ionization potential, electron affinity, electronegativity, etc.\" />\n";
print "  </metadataList>\n\n";

# Now write 'em out
my $atomicNum = 0;
for(@elements) {
    print "  <atom id=\"$elements[$atomicNum]{symbol}\">\n";
    print "    <scalar dataType=\"xsd:Integer\" dictRef=\"bo:atomicNumber\">$atomicNum</scalar>\n";

    # symbols and names
    print "    <label dictRef=\"bo:symbol\" value=\"$elements[$atomicNum]{symbol}\" />\n";
    print "    <label dictRef=\"bo:name\" xml:lang=\"en\" value=\"$elements[$atomicNum]{name}\" />\n";

    # masses and exact masses
    if ( exists $elements[$atomicNum]{mass} ) {
        print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:mass\" units=\"units:atmass\"";
        if (exists $elements[$atomicNum]{massError}) {
            print " errorValue=\"$elements[$atomicNum]{massError}\"";
        }
        print ">$elements[$atomicNum]{mass}</scalar>\n";
    }

    if ( exists $elements[$atomicNum]{exactMass} ) {
        print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:exactMass\" units=\"units:atmass\"";
        if (exists $elements[$atomicNum]{exactMassError}) {
            print " errorValue=\"$elements[$atomicNum]{exactMassError}\"";
        }
        print ">$elements[$atomicNum]{exactMass}</scalar>\n";
    }

    #Density
#    if (exists $elements[$atomicNum]{density}){
#	    print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:density\"";
#           if ( $elements[$atomicNum]{aggregation} eq "g") {
#		    print " units=\"g.l-1\"";
#	    }
#	    else{
#		    print " units=\"g.cm-3\"";
#	    }
#
#     print ">$elements[$atomicNum]{density}</scalar>\n";
#    }

    # ionization energies
    if (exists $elements[$atomicNum]{ionization} ) {
	print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:ionization\" units=\"units:ev\">$elements[$atomicNum]{ionization}</scalar>\n";
    }

    # electron affinity
    if (exists $elements[$atomicNum]{affinity} ) {
	print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:electronAffinity\" units=\"units:ev\"";
	if (exists $elements[$atomicNum]{eaError}) {
	    print " errorValue=\"$elements[$atomicNum]{eaError}\"";
	}
	print ">$elements[$atomicNum]{affinity}</scalar>\n";
    }

    # electronegativity (Pauling)
    if (exists $elements[$atomicNum]{pauling} ) {
	print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:electronegativityPauling\" units=\"boUnits:paulingScaleUnit\">$elements[$atomicNum]{pauling}</scalar>\n";
    }

	# ethymology
	if (exists $elements[$atomicNum]{nameOrigin} ) {
		print "    <scalar dataType=\"xsd:string\" dictRef=\"bo:nameOrigin\" xml:lang=\"en\">$elements[$atomicNum]{nameOrigin}</scalar>\n";
	}

    # covalent radii
    if (exists $elements[$atomicNum]{covalent} ) {
	print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:radiusCovalent\" units=\"units:ang\">$elements[$atomicNum]{covalent}</scalar>\n";
    }

    # vdW radii
    if (exists $elements[$atomicNum]{vdw} ) {
	print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:radiusVDW\" units=\"units:ang\">$elements[$atomicNum]{vdw}</scalar>\n";
    }

    # colors
    print "    <array title=\"color\" dictRef=\"bo:elementColor\" size=\"3\" dataType=\"xsd:float\">$elements[$atomicNum]{red} $elements[$atomicNum]{green} $elements[$atomicNum]{blue}</array>\n";

    # Boilingpoint
    if (exists $elements[$atomicNum]{boiling} ) {
        print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:boilingpoint\"  units=\"siUnits:kelvin\">$elements[$atomicNum]{boiling}</scalar>\n";
    }

    # Meltingpoint
    if (exists $elements[$atomicNum]{melting} ) {
        print "    <scalar dataType=\"xsd:float\" dictRef=\"bo:meltingpoint\"  units=\"siUnits:kelvin\">$elements[$atomicNum]{melting}</scalar>\n";
    }

    # Block 
    if (exists $elements[$atomicNum]{block} ) {
        print "    <scalar dataType=\"xsd:string\" dictRef=\"bo:periodTableBlock\">$elements[$atomicNum]{block}</scalar>\n";
    }
    
    # The country where the element has been discovered
    if (exists $elements[$atomicNum]{country} ) {
      print "    <array dataType=\"xsd:string\""; 

      my @count = split(',', $elements[$atomicNum]{country});
      # if there are more than one discoverer-countries. So this scripts has to
      # print the delim=";" size="2" xml as well. The size is the
      # number of names
      if (scalar(@count) > 1){
        print " delimiter=\",\" size=\"". scalar(@count) ."\"";
      }

      print " dictRef=\"bo:discoveryCountry\">$elements[$atomicNum]{country}</array>\n";
    }

    # Discovery 
    if (exists $elements[$atomicNum]{discoveryDate} ) {
        print "    <scalar dataType=\"xsd:date\" dictRef=\"bo:discoveryDate\">$elements[$atomicNum]{discoveryDate}</scalar>\n";
    }
    if (exists $elements[$atomicNum]{discoverers} ) {
        print "    <array dataType=\"xsd:string\"";
		my @count = split(';', $elements[$atomicNum]{discoverers});
		# if there are more than one discoverer. So this scripts has to
		# print the delim=";" size="2" xml as well. The size is the
		# number of names
		if (scalar(@count) > 1){
			print " delimiter=\";\" size=\"". scalar(@count) ."\"";
		}
		
		print " dictRef=\"bo:discoverers\">$elements[$atomicNum]{discoverers}</array>\n";
    }
    
	# Periods
    if (exists $elements[$atomicNum]{period} ) {
        print "    <scalar dataType=\"xsd:int\" dictRef=\"bo:period\">$elements[$atomicNum]{period}</scalar>\n";
    }
	
	# Acidicbehaviour
    if (exists $elements[$atomicNum]{acidicbehaviour} ) {
        print "    <scalar dataType=\"xsd:int\" dictRef=\"bo:acidicbehaviour\">$elements[$atomicNum]{acidicbehaviour}</scalar>\n";
    }
	
	# Group
    if (exists $elements[$atomicNum]{group} ) {
        print "    <scalar dataType=\"xsd:int\" dictRef=\"bo:group\">$elements[$atomicNum]{group}</scalar>\n";
    }
	
	# Electronic configuration
    if (exists $elements[$atomicNum]{electronicConfiguration} ) {
        print "    <scalar dataType=\"xsd:string\" dictRef=\"bo:electronicConfiguration\">$elements[$atomicNum]{electronicConfiguration}</scalar>\n";
    }
	
	# Family
    if (exists $elements[$atomicNum]{family} ) {
        print "    <scalar dataType=\"xsd:string\" dictRef=\"bo:family\">$elements[$atomicNum]{family}</scalar>\n";
    }

    print "  </atom>\n";
    $atomicNum++;
}

print "</list>\n";

exit(0);

sub readColors {
    die "Can't find colors.txt!\n" if (! -e "colors.txt");

    open COLORS, "<colors.txt";
    while (<COLORS>) {
	chomp;
	
	next if ($_ =~ m/^#/);
	my @line = split; # let's just save the line items
	my $atomicNum = shift @line;
	
	$elements[$atomicNum]{red} = shift @line;
	$elements[$atomicNum]{green} = shift @line;
	$elements[$atomicNum]{blue} = shift @line;
    }
    close COLORS;
}

sub readNamesAndSymbols {
    die "Can't find names.txt!\n" if (! -e "names.txt");

    open NAMES, "<names.txt";
    while (<NAMES>) {
	chomp;
	
	next if ($_ =~ m/^#/);
	my @line = split;
	my $atomicNum = shift @line;
	
	$elements[$atomicNum]{symbol} = shift @line;
	$elements[$atomicNum]{name} = shift @line;
    }
    close NAMES;
}


sub readAtomicMasses {
    die "Can't find mass.txt!\n" if (! -e "mass.txt");

open MASS, "<mass.txt";
while (<MASS>) {
    chomp;

    next if ($_ =~ m/^#/);
    my @line = split;
    my $atomicNum = shift @line;

    shift @line; # symbol
    $elements[$atomicNum]{mass} = shift @line;
    if (my $massError = shift @line) {
	$elements[$atomicNum]{massError} = $massError;
    }
}
close MASS;
}

sub readExactMasses {
    die "Can't find exact-masses.txt!\n" if (! -e "exact-masses.txt");

    open EXACTMASS, "<exact-masses.txt";
    while (<EXACTMASS>) {
	chomp;
	
	next if ($_ =~ m/^#/);
	my @line = split;
	my $atomicNum = shift @line;
	
	shift @line; # most common isotope
	$elements[$atomicNum]{exactMass} = shift @line;
	if (my $massError = shift @line) {
	    $elements[$atomicNum]{exactMassError} = $massError;
	}
    }
    close EXACTMASS;
}

sub readIonization {
    die "Can't find ionization.txt!\n" if (! -e "ionization.txt");

    open IONIZE, "<ionization.txt";
    while (<IONIZE>) {
	chomp;
	
	next if ($_ =~ m/^#/);
	my @line = split;
	my $atomicNum = shift @line;
	
	shift @line; # symbol
	$elements[$atomicNum]{ionization} = shift @line;
    }
    close IONIZE;
}

sub readElectronAffinity {
    die "Can't find electron-affinity.txt!\n" if (! -e "electron-affinity.txt");

    open EA, "<electron-affinity.txt";
    while (<EA>) {
	chomp;
	
	next if ($_ =~ m/^#/);
	my @line = split;
	my $atomicNum = shift @line;
	
	$elements[$atomicNum]{affinity} = shift @line;
	if (my $eaError = shift @line) {
	    $elements[$atomicNum]{eaError} = $eaError;
	}
    }
    close EA;
}

sub readPauling {
    die "Can't find electroneg-pauling.txt!\n" if (! -e "electroneg-pauling.txt");

    open PAULING, "<electroneg-pauling.txt";
    while (<PAULING>) {
	chomp;
	
	next if ($_ =~ m/^#/);
	my @line = split;
	my $atomicNum = shift @line;
	
	$elements[$atomicNum]{pauling} = shift @line;
    }
    close PAULING;
}

sub readEthymology {
	die "Can't find nameorigin.txt!\n" if (! -e "nameorigin.txt");

	open ORIGIN, "<nameorigin.txt";
	while (<ORIGIN>) {
		chomp;

		next if ($_ =~ m/^#/);
		if ($_ =~ m/^(\d*)\s*(.*)/) {
			$elements[$1]{nameOrigin} = $2;
		}
	}
	close ORIGIN;
}

sub readRadiiCovalent {
    die "Can't find radii-covalent.txt!\n" if (! -e "radii-covalent.txt");

    open COVALENT, "<radii-covalent.txt";
    while (<COVALENT>) {
	chomp;
	
	next if ($_ =~ m/^#/);
	my @line = split;
	my $atomicNum = shift @line;
	
	$elements[$atomicNum]{covalent} = shift @line;
    }
    close COVALENT;
}

sub readRadiiVanDerWaals {
    die "Can't find radii-vdw.txt!\n" if (! -e "radii-vdw.txt");

    open VDW, "<radii-vdw.txt";
    while (<VDW>) {
	chomp;
	
	next if ($_ =~ m/^#/);
	my @line = split;
	my $atomicNum = shift @line;
	
	$elements[$atomicNum]{vdw} = shift @line;
    }
    close VDW;
}

sub readBoilingpoint{
    die "Can't find boilingpoint.txt!\n" if (! -e "boilingpoint.txt");

    open BOILING, "<boilingpoint.txt";
    while (<BOILING>) {
        chomp;

	next if ($_ =~ m/^#/);
        my @line = split;
        my $atomicNum = shift @line;

        $elements[$atomicNum]{boiling} = shift @line;
    }
    close BOILING;
}

sub readMeltingpoint{
    die "Can't find meltingpoint.txt!\n" if (! -e "meltingpoint.txt");

    open MELTING, "<meltingpoint.txt";
    while (<MELTING>) {
        chomp;

	next if ($_ =~ m/^#/);
        my @line = split;
        my $atomicNum = shift @line;

        $elements[$atomicNum]{melting} = shift @line;
    }
    close MELTING;
}

sub readBlocks{
    die "Can't find blocks.txt!\n" if (! -e "blocks.txt");

    open BLOCKS, "<blocks.txt";
    while (<BLOCKS>) {
        chomp;

	next if ($_ =~ m/^#/);
        my @line = split;
        my $atomicNum = shift @line;

        $elements[$atomicNum]{block} = shift @line;
    }
    close BLOCKS;
}

sub readDatesAndDiscoveres {
    die "Can't find dates.txt!\n" if (! -e "dates.txt");

	open DATES, "<dates.txt";
	while (<DATES>){
		chomp;
		my @data = split(':', $_);

		next if ($_ =~ m/^#/);
		if (scalar(@data) == 2) {
			my $atomicNum = $data[0];
			$elements[$atomicNum]{discoveryDate} = $data[1];
		}
		elsif (scalar(@data) == 3) {
			my $atomicNum = $data[0];
			$elements[$atomicNum]{discoveryDate} = $data[1];
			$elements[$atomicNum]{discoverers} = $data[2];
		}
	}
	close DATES;
}

sub readPeriods{
    die "Can't find periods.txt!\n" if (! -e "periods.txt");

    open PERIODS, "<periods.txt";
    while (<PERIODS>) {
        chomp;

	next if ($_ =~ m/^#/);
        my @line = split;
        my $atomicNum = shift @line;

        $elements[$atomicNum]{period} = shift @line;
    }
    close PERIODS;
}


sub readElectronicConfiguration {
    die "Can't find electronic-configuration.txt!\n" if (! -e "electronic-configuration.txt");

	open ECONF, "<electronic-configuration.txt";
	while (<ECONF>) {
		chomp;

		next if ($_ =~ m/^#/);
		if ($_ =~ m/^(\d*)\s*(.*)/) {
			$elements[$1]{electronicConfiguration} = $2;
		}
	}
	close ECONF;
}

sub readAcidicbehaviour{
    die "Can't find acidicbehaviour.txt!\n" if (! -e "acidicbehaviour.txt");

    open PERIODS, "<acidicbehaviour.txt";
    while (<PERIODS>) {
        chomp;
		
	next if ($_ =~ m/^#/);
        my @line = split;
        my $atomicNum = shift @line;

        $elements[$atomicNum]{acidicbehaviour} = shift @line;
    }
    close PERIODS;
}
sub readGroups{
    die "Can't find groups.txt!\n" if (! -e "groups.txt");

    open GROUPS, "<groups.txt";
    while (<GROUPS>) {
        chomp;

	next if ($_ =~ m/^#/);
        my @line = split;
        my $atomicNum = shift @line;

        $elements[$atomicNum]{group} = shift @line;
    }
    close GROUPS;
}

sub readFamilies{
    die "Can't find family.txt!\n" if (! -e "family.txt");

    open FAMILIES, "<family.txt";
    while (<FAMILIES>) {
        chomp;

	next if ($_ =~ m/^#/);
        my @line = split;
        my $atomicNum = shift @line;

        $elements[$atomicNum]{family} = shift @line;
    }
    close FAMILIES;
}

sub readDensity{
    die "Can't find density.txt!\n" if (! -e "density.txt");

  open DENSITY, "<density.txt";
  while (<DENSITY>) {
    chomp;

    next if ($_ =~ m/^#/);
    my @line = split;
    my ($atomicNum, $density, $aggr, @restOfLine) = split;

    $elements[$atomicNum]{density} = $density;
    $elements[$atomicNum]{aggregation} = $aggr;
  }
  close DENSITY;
}

sub readCountries{
    die "Can't find discovery_countries.txt!\n" if (! -e "discovery_countries.txt");

    open COUNTRIES, "<discovery_countries.txt";
    while (<COUNTRIES>) {
        chomp;

	next if ($_ =~ m/^#/);
        my @line = split;
        my $atomicNum = shift @line;

        $elements[$atomicNum]{country} = shift @line;
    }
    close COUNTRIES;
}


bodr/scripts/crystalstructure.rb
#!/usr/bin/env ruby
# (c) 2006 Carsten Niehaus <cniehaus@kde.org>
# License: MIT

#Example-syntax of a CML 2.3-crystal
#<molecule id="crystal1">
#  <crystal z="4">
#    <scalar id="sc1" title="a" units="units:angstrom">4.500</scalar>
#    <scalar id="sc2" title="b" units="units:angstrom">4.500</scalar>
#    <scalar id="sc3" title="c" units="units:angstrom">4.500</scalar>
#    <scalar id="sc4" title="alpha" units="units:degrees">90</scalar>
#    <scalar id="sc5" title="beta" units="units:degrees">90</scalar>
#    <scalar id="sc6" title="gamma" units="units:degrees">90</scalar>
#    <symmetry id="s1" spaceGroup="Fm3m"/>
#  </crystal>
#  <atomArray>
#    <atom id="a1" elementType="Na" formalCharge="1" xFract="0.0" yFract="0.0" zFract="0.0"/>
#    <atom id="a2" elementType="Cl" formalCharge="-1" xFract="0.5" yFract="0.0" zFract="0.0"/>
#  </atomArray>
#</molecule>

class Crystal
  #Check if the crystal is valid or not
  def valid?
    # 1-2   Triclinic
    # 3-15  Monoclinic
    # 16-74 Orthorhombic
    # 75-142  Tetragonal
    # 143-167 Rhombohedral (Trigonal)
    # 168-194 Hexagonal
    # 195-230 Cubic

    if not @spacegroup <= 230
      puts "Spacegroup to high: "+@spacegroup.to_s
      return false
    elsif @spacegroup < 1
      puts "Spacegroup to small: "+@spacegroup.to_s
      return false
    end

    case @spacegroup
    when 1..2
      if @system != "tri"
        puts "Triclinic (1-2): #{@system}, #{@spacegroup.to_s}"
      end
      if @a == @b or @a == c or @b == @c
        puts "Same length! #{@a}, #{@b}, #{@c}"
      end
      if @alpha == @beta or @alpha == @gamma or @beta == @gamma
        puts "Same angle! #{@alpha}, #{@beta}, #{@gamma}"
      end
    when 3..15
      if @system != "mono"
        puts "#{@id}: Monoclinic (3-15): #{@system}, #{@spacegroup.to_s}"
      end
      if @a == @b or @a == c or @b == @c
        puts "Same length! #{@a}, #{@b}, #{@c}"
      end
    when 16..74
      if @system != "or"
        puts "#{@id}: Orthorhombic (16-74): #{@system}, #{@spacegroup.to_s}"
      end
      if @a == @b or @a == c or @b == @c
        puts "Same length! #{@a}, #{@b}, #{@c}"
      end
    when 75..142
      if @system != "tet"
        puts "#{@id}: Tetragonal (75-142): #{@system}, #{@spacegroup.to_s}"
      end
      if @a != @b 
        puts "#{@a} != #{@b}"
      end
    when 143..167
      if @system != "rh"
        puts "#{@id}: Rhombohedral (143-167): #{@system}, #{@spacegroup.to_s}"
      end
    when 168..194
      if @system != "hcp"
        puts "#{@id}: Hexagonal (168-194): #{@system}, #{@spacegroup.to_s}"
      end
    when 195..130
      if @system != "ccp" and @system != "bcc" and @system != "sc"
        puts "#{@id}: Cubic (195-230): #{@system}, #{@spacegroup.to_s}"
      end
    end

    return true
  end

  attr_accessor :a, :b, :c, :alpha, :beta, :gamma, :spacegroup, :id, :system
end

$xmlfile = <<EOS
<?xml version="1.0" encoding="utf-8"?>
<list xmlns="http://www.xml-cml.org/schema"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xmlns:bo="http://www.blueobelisk.org/dict/terminology"
      xmlns:boUnits="http://www.blueobelisk.org/dict/units"
      xmlns:units="http://www.xml-cml.org/units/units"
      xmlns:siUnits="http://www.xml-cml.org/units/siUnits"
      xmlns:bibx="http://bibtexml.sf.net/"
      xsi:schemaLocation=\"http://www.xml-cml.org/schema ../schemas/cml25.xsd
                         http://bibtexml.sf.net/        ../schemas/bibtexml.xsd"
      id="crystalStructure"
      title="properties of the crystals">

  <metadataList>
    <metadata name="dc:title" content="Blue Obelisk Element Repository" />
    <metadata name="dc:creator" content="The Blue Obelisk Movement" />
    <metadata name="dc:license" content="The MIT License" />
    <metadata name="dc:contributor" content=\"Carsten Niehaus" />
  </metadataList>
EOS

# return the chemical symbol of the element with the given number
def get_symbol(number)
  elementsymbols = ["dummy","H", "He", "Li", "Be", "B", "C", "N", "O", "F", "Ne", "Na", "Mg", "Al", "Si", "P", "S", "Cl", "Ar", "K", "Ca", "Sc", "Ti", "V", "Cr", "Mn", "Fe", "Co", "Ni", "Cu", "Zn", "Ga", "Ge", "As", "Se", "Br", "Kr", "Rb", "Sr", "Y", "Zr", "Nb", "Mo", "Tc", "Ru", "Rh", "Pd", "Ag", "Cd", "In", "Sn", "Sb", "Te", "I", "Xe", "Cs", "Ba", "La", "Ce", "Pr", "Nd", "Pm", "Sm", "Eu", "Gd", "Tb", "Dy", "Ho", "Er", "Tm", "Yb", "Lu", "Hf", "Ta", "W", "Re", "Os", "Ir", "Pt", "Au", "Hg", "Tl", "Pb", "Bi", "Po", "At", "Rn", "Fr", "Ra", "Ac", "Th", "Pa", "U", "Np", "Pu", "Am", "Cm", "Bk", "Cf", "Es", "Fm", "Md", "No", "Lr", "Rf", "Db", "Sg", "Bh", "Hs", "Mt", "Ds", "Rg", "Uub", "Uut", "Uuq", "Uup", "Uuh"]

  return elementsymbols[number]
end

def process_element(content)
  stuff = content.split(' ')

  crystal = Crystal.new

  crystal.id = stuff[0]
  crystal.a = stuff[3].to_f
  crystal.b = stuff[4].to_f
  crystal.c = stuff[5].to_f
  crystal.alpha = stuff[6].to_f
  crystal.beta = stuff[7].to_f
  crystal.gamma = stuff[8].to_f
  crystal.spacegroup = stuff[2].to_i
  crystal.system = stuff[1]

  symbol = get_symbol(stuff[0].to_i)

  return unless crystal.valid?

xml = <<ELEMENTXML
  <molecule id="#{symbol}#{crystal.id}">
    <crystal>
      <scalar id="sc1" title="a" units="units:pm">#{crystal.a}</scalar>
      <scalar id="sc2" title="b" units="units:pm">#{crystal.b}</scalar>
      <scalar id="sc3" title="c" units="units:pm">#{crystal.c}</scalar>
      <scalar id="sc4" title="alpha" units="units:degrees">#{crystal.alpha}</scalar>
      <scalar id="sc5" title="beta" units="units:degrees">#{crystal.beta}</scalar>
      <scalar id="sc6" title="gamma" units="units:degrees">#{crystal.gamma}</scalar>
      <symmetry id="s1" spaceGroup="#{crystal.spacegroup}" />
    </crystal>
    <atomArray>
      <atom id="a1" elementType="#{symbol}" xFract="0.0" yFract="0.0" zFract="0.0" />
    </atomArray>
  </molecule>
ELEMENTXML

  $xmlfile << xml
end

File.open("crystalstructures.txt") do |file|
  file.each {|line| 
    process_element(line) unless line =~ /^#/
  }
end

$xmlfile << "</list>\n";

File.open("crystalstructures.xml", "w+") { |f| 
  f << $xmlfile 
}
